examples/complex.dx

```diff
@@ -0,0 +1,227 @@
+'## Complex Numbers
+
+
+'## Things to add to prelude
+
+def sign (x:Float) : Float =
+  case x > 0.0 of
+    True -> 1.0
+    False -> case x < 0.0 of
+      True -> -1.0
+      False -> x
+
+def copysign (a:Float) (b:Float) : Float =
+  case b > 0.0 of
+    True -> a
+    False -> case b < 0.0 of
+      True -> (-a)
+      False -> 0.0
+
+def atan_neg1_to1 (x:Float) : Float = 
+  -- From https://arxiv.org/abs/1508.03211
+  -- Only accurate in the range [-1, 1]
+  s = x * x
+  r = 0.0027856871
+  r = r * s - 0.0158660002
+  r = r * s + 0.042472221
+  r = r * s - 0.0749753043
+  r = r * s + 0.106448799
+  r = r * s - 0.142070308
+  r = r * s + 0.199934542
+  r = r * s - 0.333331466
+  r = r * s
+  r * x + x
+
+def atan2 (y:Float) (x:Float) : Float =
+  -- Based off of the Tensorflow implementation at
+  -- https://github.com/tensorflow/mlir-hlo/blob/master/lib/Dialect/mhlo/transforms/legalize_trigonometric_to_approximation.cc#L147
+  -- todo: Handle nans and infs
+  ax = abs x
+  ay = abs y
+  zero_to_one = (min ax ay) / (max ax ay)
+  a = atan_neg1_to1 zero_to_one
+  a = select (ax <= ay) ((pi / 2.0) -a) a
+  a = select (x < 0.0) pi a
+  t = select (x < 0.0) pi 0.0
+  a = select (y == 0.0) t a
+  copysign a y
+
+def atan (x:Float) : Float = atan2 x 1.0
+
+-- Todo: better numerics for very large and small values.
+def sinh (x:Float) : Float = ((exp x) - exp (-x)) / 2.0
+def cosh (x:Float) : Float = ((exp x) + exp (-x)) / 2.0
+def tanh (x:Float) : Float = (1.0 - (exp (-x))) / (1.0 + (exp (-x)))
+
+
+'## Complex number definition
+
+-- Maybe use a newtype or typeclass, to avoid accidentally
+-- treating a tuple of two floats as a complex number?
+Complex = (Float & Float)  -- real, imaginary
+def real (x : Complex) : Float = fst x
+def imag (x : Complex) : Float = snd x
+
+@instance allCloseComplex : HasAllClose Complex =
+  MkHasAllClose \atol rtol (a,b):Complex (c,d):Complex.
+    fstclose = abs (a - c) <= ((fst atol) + (fst rtol) * abs c)
+    sndclose = abs (b - d) <= ((snd atol) + (snd rtol) * abs d)
+    fstclose && sndclose
+
+@instance defaultToleranceComplex : HasDefaultTolerance Complex =
+  MkHasDefaultTolerance (FToF32 0.00001, FToF32 0.00001) (FToF32 0.0001, FToF32 0.00001)
```

Comment:
It would be good to use the values derived from `Float`'s `HasDefaultTolerance` instance here, so that they change whenever we update the prelude.

Reply:
I agree.  I think I can do this if I make Complex a typeclass that takes in a Floating point precision, although I haven't tried to do that yet.

Reply:
Okay, I figured out how to derive the values from `Float`'s `HasDefaultTolerance`.